### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by rewong03**
CSV, Postgres, and Parquet serializers

### CODE REVIEW

File: src/main/java/dgen/column/ColumnConfig.java [hunk 1]
```diff
@@ -20,6 +20,9 @@
     public static final String COLUMN_ID = "column.id";
     private static final String COLUMN_ID_DOC = "Indicates unique ID of column";
 
+    public static final String COLUMN_TYPE = "column.type";
+    private static final String COLUMN_TYPE_DOC = "Indicates type of column";
```
Comment (raulcf) [inline src/main/java/dgen/column/ColumnConfig.java]: Does type mean 'data type'?

File: src/main/java/dgen/column/ColumnGenerator.java [hunk 2]
```diff
@@ -47,10 +48,10 @@ public ColumnGenerator(ColumnConfig columnConfig) {
             this.ang = new RandomAttributeNameGenerator(rnd);
         }
 
-        if (columnConfig.getObject("datatype") == null) {
+        if (columnConfig.getObject("column.type") == SpecType.DEFFOREIGNKEY) {
```
Comment (raulcf) [inline src/main/java/dgen/column/ColumnGenerator.java:51]: It does not mean 'data type' based on this line, so probably good to explain what that 'type' above means.

File: src/main/java/dgen/utils/serialization/ParquetSerializer.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                .withCompressionCodec(CompressionCodecName.SNAPPY)
+                .build()) {
+
+            for (GenericData.Record record : recordsToWrite) {
+                writer.write(record);
+            }
+        }
+    }
+
+    public void readFromParquet(Path filePathToRead) throws IOException {
```
Comment (raulcf) [inline src/main/java/dgen/utils/serialization/ParquetSerializer.java:176]: is this method for testing purposes?

File: src/main/java/dgen/utils/serialization/PostgresSerializer.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            statement.append("\n");
+        }
+
+        statement.delete(statement.length() - 6, statement.length());
+        statement.append(";");
+
+        return statement.toString();
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (raulcf) [inline src/main/java/dgen/utils/serialization/PostgresSerializer.java:211]: have you tested loading this into a postgres database? does the resulting schema look like what it should?

**Review by raulcf — COMMENTED** [review]
Is the output serializer indicated in the high-level spec? What about the output directory for the dataset and the metadata?

I left a few comments, mostly related to testing of the serializers.

**Comment by rewong03** [comment]
Right now users don't specify any serialization within the high-level spec. They choose the desired serializer by creating a `Serializer` object and passing the dataset and output paths. Should this be changed so that the user specifies the serializer in the spec and calling the `.generate()` method on the dataset automatically serializes?

File: src/main/java/dgen/utils/serialization/PostgresSerializer.java (hunk 4 above)
Comment (rewong03) [reply src/main/java/dgen/utils/serialization/PostgresSerializer.java:211]: Yes, it creates a new schema within the current database and loads the tabes, data, relationships, etc. properly

File: src/main/java/dgen/column/ColumnGenerator.java (hunk 2 above)
Comment (rewong03) [reply src/main/java/dgen/column/ColumnGenerator.java:51]: Fixed in 4ec4ff75482108855ea85906f3e59ab4525f0b98

File: src/main/java/dgen/utils/serialization/ParquetSerializer.java (hunk 3 above)
Comment (rewong03) [reply src/main/java/dgen/utils/serialization/ParquetSerializer.java:176]: It's for testing but I'll leave it in for now. I added a short comment that says its for testing in 4ec4ff75482108855ea85906f3e59ab4525f0b98

**Comment by raulcf** [comment]
> Right now users don't specify any serialization within the high-level spec. They choose the desired serializer by creating a Serializer object and passing the dataset and output paths. Should this be changed so that the user specifies the serializer in the spec and calling the .generate() method on the dataset automatically serializes?

When you say 'passing the dataset' what does that mean? how is the dataset represented when passed to the serializer?

**Comment by rewong03** [comment]
The dataset is just a java object that you pass as a parameter to a serializer's `.serialize()` method.

File: src/main/java/dgen/utils/serialization/PostgresSerializer.java (hunk 4 above)
Comment (raulcf) [reply src/main/java/dgen/utils/serialization/PostgresSerializer.java:211]: Alright. Let me merge this PR and we will discuss the user interface later.

**Pull request merged**